File: 1701-1800/1741-1750/1750M_Minimum_Length_of_String_After_Deleting_Similar_Ends/java/main.java
// 1750. Minimum Length of String After Deleting Similar Ends
// https://leetcode.com/problems/minimum-length-of-string-after-deleting-similar-ends/description/
// T.C.: O(n)
// S.C.: O(1)

// java main.java
class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();

        String s = "ca";
        int expected = 2;
        int result = solution.minimumLength(s);
        System.out.println(result + " " + (expected == result ? "true" : "false"));

        s = "cabaabac";
        expected = 0;
        result = solution.minimumLength(s);
        System.out.println(result + " " + (expected == result ? "true" : "false"));

        s = "aabccabba";
        expected = 3;
        result = solution.minimumLength(s);
        System.out.println(result + " " + (expected == result ? "true" : "false"));

        s = "bbbbbbbbbbbbbbbbbbbbbbbbbbbabbbbbbbbbbbbbbbccbcbcbccbbabbb";
        expected = 1;
        result = solution.minimumLength(s);
        System.out.println(result + " " + (expected == result ? "true" : "false"));

        s = "abbbbbbbbbbbbbbbbbbba";
        expected = 0;
        result = solution.minimumLength(s);
        System.out.println(result + " " + (expected == result ? "true" : "false"));
    }
}

class Solution {
    public int minimumLength(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right && s.charAt(left) == s.charAt(right)) {
            char c = s.charAt(left);
            while (left <= right && s.charAt(left) == c) {
                left++;
            }
            while (right > left && s.charAt(right) == c) {
                right--;
            }
        }

        return right - left + 1;
    }
}

File: 1701-1800/1741-1750/1750M_Minimum_Length_of_String_After_Deleting_Similar_Ends/rust/src/main.rs
// 1750. Minimum Length of String After Deleting Similar Ends
// https://leetcode.com/problems/minimum-length-of-string-after-deleting-similar-ends/description/
// T.C.: O(n)
// S.C.: O(1)
use colored::Colorize;
use colored::ColoredString;

struct Solution;

impl Solution {
    pub fn minimum_length(s: String) -> i32 {
        let s = s.as_bytes();
        let mut left = 0;
        let mut right = s.len() - 1;
        while left < right && s[left] == s[right] {
            let ch = s[left];
            while left <= right && s[left] == ch {
                left += 1;
            }

            while left < right && s[right] == ch {
                right -= 1;
            }
        }

        return (right + 1 - left) as i32;
    }
}

fn main() {
    let mut s = String::from("ca");
    test(s, 2);

    s = String::from("cabaabac");
    test(s, 0);

    s = String::from("aabccabba");
    test(s, 3);

    s = String::from("bbbbbbbbbbbbbbbbbbbbbbbbbbbabbbbbbbbbbbbbbbccbcbcbccbbabbb");
    test(s, 1);

    s = String::from("abbbbbbbbbbbbbbbbbbba");
    test(s, 0);
}

fn test(s: String, expected: i32) {
    let result = Solution::minimum_length(s);

    println!("\nresult: {}\nassert: {}", number_colored(result), bool_colored(result == expected));
}

fn bool_colored(value: bool) -> ColoredString {
    let text;
    if value {
        text = value.to_string().green();
    } else {
        text = value.to_string().red();
    }
    return text;
}

fn number_colored(value: i32) -> ColoredString {
    return value.to_string().yellow();
}
